projects/media-viewer/src/lib/toolbar/media-viewer-toolbar/media-viewer-toolbar.component.ts

```diff
@@ -0,0 +1,281 @@
+    this.cdr.detectChanges();
+
+    for (const element of toolbarElements) {
+      this.totalButtonSpace += element.getBoundingClientRect().width;
+    }
+
+    // add the main toolbar left margin to the available space
+    this.totalButtonSpace = this.totalButtonSpace + this.mvToolbarMain.nativeElement.getBoundingClientRect().left;
+
+    // Loop over all elements and set sum of widths for each toolbar element
+    for (const element of toolbarElements) {
+      // Use bounding client rectangle width for accuracy
+      totalSpace += element.getBoundingClientRect().width;
+      this.breakWidths.push(totalSpace);
+    }
+    // Call calculation method
+    this.checkCalculation();
+  }
+
+  @HostListener('window:resize', [])
+  public onResize() {
+    this.cdr.detectChanges();
+    // Call calculation method
+    this.checkCalculation();
+  }
+
+  private checkCalculation() {
+    // if the more options button is visible and we have enough space to all the buttons then show them all
+    if (this.totalButtonSpace < this.mvMenuItems.nativeElement.getBoundingClientRect().width) {
+      for (const buttonState of this.toolbarButtonVisibilityStates) {
+        buttonState.showOnToolbar = true;
+      }
+      return;
+    }
+
+    // Get current space of main toolbar element, including the right margin (if any) to the "More options" button
+    const availableSpace = this.mvToolbarMain.nativeElement.getBoundingClientRect().width +
+      parseFloat(getComputedStyle(this.mvToolbarMain.nativeElement).marginRight);
+    // Get space required for all elements in main toolbar element (i.e. break-width of the last visible element from breakWidth array)
+    const indexOfLastVisibleElement = this.getIndexOfLastVisibleToolbarButton();
+    const totalSpaceRequired = this.breakWidths[indexOfLastVisibleElement];
+
+    // If more space is needed than is available, hide last element
+    if (totalSpaceRequired > availableSpace) {
+      // Mark the toolbar button as hidden in the button states array
+      this.setToolbarButtonVisible(this.toolbarButtonVisibilityStates[indexOfLastVisibleElement].id, false);
+
+      // Apply changes
+      this.cdr.detectChanges();
+
+      // Call this method to recalculate for rest of toolbar elements
+      this.checkCalculation();
+    } else if (this.toolbarHasHiddenButtons() && this.breakWidths[indexOfLastVisibleElement + 1] <= availableSpace) {
+      // If there are some toolbar buttons marked as hidden in the button states array, AND there is room for the first of these,
+      // mark the button as visible
+      for (const buttonState of this.toolbarButtonVisibilityStates) {
+        if (!buttonState.showOnToolbar) {
+          buttonState.showOnToolbar = true;
+          break;
+        }
+      }
+
+      // Call this method to recalculate for rest of toolbar elements
+      this.checkCalculation();
+    }
+  }
+
+  public isToolbarButtonVisible(buttonId: string): boolean {
```

Comment:
if we're gonna use this, it should be a map rather than an array, so that we can go directly to the button we want